New info about, Why is good this template to work with defined architectures (#89)
moove-it/react-native-template: README.md

## Usage

#### Use Template button

Click the "Use this template" button above the file list, then use the Owner drop-down menu, and select the account you want to own the repository. Creating a repository from a template has the following advantages:

- A repository created from a template starts with a single commit.
- Commits to a repository created from a template do appear in your contribution graph.
- Creating a repository from a template starts a new project quickly.

### Option 1: Using React-Native-Rename

You can start by cloning this repository and using [react-native-rename](https://github.com/junedomingo/react-native-rename). In the current state of this project, it should give you no issues at all, just run the script, delete your node modules and reinstall them and you should be good to go.

Keep in mind that this library can cause trouble if you are renaming a project that uses `Pods` on the iOS side.

After that you should proceed as with any javascript project:

- Go to your project's root folder and run `npm install`.
- If you are using xcode 12.5 or higher got to /ios and execute `pod install --repo-update`

> - Run `npm run ios:development` or `npm run android:development` to start your application!
> 
> (Using yarn: `yarn ios:development` or `yarn android:development`)

Reviewer: ```suggestion
- Run `npm run ios` or `npm run android` to start your application!

(Using yarn: `yarn ios` or `yarn android`)
```

these scripts do not exist. the available options are:

- android
- android:prod
- android:staging

- ios
- ios:prod
- ios:staging

- Give it a proper name on the top left. For instance: "qa"
- Then edit the newly created scheme to make it use a different env file. From the same "manage scheme" window:

Expand the "Build" settings on left

- Click "Pre-actions", and under the plus sign select "New Run Script Action"
- Where it says "Type a script or drag a script file", type: `echo ".env.qa" > /tmp/envfile` replacing `.env.qa` with your file.

## Styleguide

For coding styling we decided to go with ESLint and [React Native community's styleguide](https://github.com/facebook/react-native/tree/master/packages/eslint-config-react-native-community#readme).

# How to use it

The idea of this section is to explain how the template composition is better and easier to use when you try to use well formed arquitectures, especially using redux flow.

First keep in mind that you will see the same way of classification in multiple places of the app. Different resources orchestrated by an index.js that allow you to call it as a single instance. This structure lets you use a simple way to import things.

With that in mind we are going to look at each folder to explain how to use it.

## Components

> The components are the basic structure of an application in react native, but specifically for this template the idea of ​​keeping the complexity of the development to a minimum, for this the ideal is that all the components are at the same nesting folder level. A component is only created if a specific view or functionality is being worked on or if the same object is to be reused in multiple parts of the application.

Reviewer: Let's try not to be dogmatic here. When projects scale, this folder always ends up with some kind of nesting or internal structure anyway.

```suggestion
Components are the basic blocks of a react native application, but since we​​ aim to minimize development complexity, all the components are at the same nesting level.
```

Another important thing is the use of propTypes to check the kind of data that your components need to work properly. If the component receives some data from others, the type of this props must be defined and in case you need it the default value of the property too.

### Static resources:

To keep an application scalable and organized, the global static resources that are used in the application have to be created in a specific file.

### We manage three main folders for that:

> - Assets: Here you can store all the images and icons that you need through the app. You have as example the icon ic_home.png, to respond with the different device screen densities just create inside the same folder the image and all the scaled versions that you need.

Reviewer: I may be wrong but I think RN only handles x1 x2 and x3, so maybe we should advise to use this scales

> - Localization: This folder contains all the language objects that you need to create a multilingual application. Create a file for each language, inside define an object and maintain the nesting sorted by the screen that contains the text that you need and the text you want to show. As the last step, remember to create a reference inside the Localization.js file and add to LocalizedStrings.

Reviewer: ```suggestion
- Localization: This folder contains all the locale objects that you need to create a multilingual application. Create a file for each locale, inside define an object and maintain the nesting sorted by the screen that contains the text. As the last step, remember to create a reference inside the Localization.js file and add it to LocalizedStrings.
```

- Theme: Here you can define all the styles that you use in different screens. To make easier the interaction of the application with device options as example you can create here assets as light and dark color palette

## Redux

Once the components are defined, they are tied to the management of information through the application. For this, Redux is implemented with the store-reducer-action structure as usual, however, not only the data is handled through the actions but the success and error responses are also defined by the same form.

The template uses hooks for communication with the redux store. In that way it has two scenarios.

- When you need to call an action you have to create an useDispatch function that lets you interact with the store.

> - To see what’s inside the store you should create a useSelector to extract specific info from the general store.

Reviewer: I would remove this since it describes how to use react-redux. Although it could be a good hint for newcomers I. think we should aim to keep the README as simple as possible.

With that in mind we are going to look at each folder to explain how to use it.

## Components

Components are the basic blocks of a react native application, but since we​​ aim to minimize development complexity, all the components are at the same nesting level.

Another important thing is the use of propTypes to check the kind of data that your components need to work properly. If the component receives some data from others, the type of this props must be defined and in case you need it the default value of the property too.

### Static resources:

To keep an application scalable and organized, the global static resources that are used in the application have to be created in a specific file.

### We manage three main folders for that:

- Assets: Here you can store all the images and icons that you need through the app. You have as example the icon ic_home.png, to respond with the different device screen densities just create inside the same folder the image and all the scaled versions that you need. RN only handles x1, x2 and x3, in this case you have.

  - assets

>     - ic_home

Reviewer: ```suggestion
```

    - ic_home.png
    - ic_home@2x.png
    - ic_home@3x.png

- Localization: This folder contains all the locale objects that you need to create a multilingual application. Create a file for each locale, inside define an object then maintain the nesting sorted by the screen that contains the text that you need and the text you want to show. As the last step, remember to create a reference inside the Localization.js file and add it to LocalizedStrings.
- Theme: Here you can define all the styles that you use in different screens. To make easier the interaction of the application with device options as example you can create here assets as light and dark color palette

## Redux

Once the components are defined, they are tied to the management of information through the application. For this, Redux is implemented with the store-reducer-action structure as usual, however, not only the data is handled through the actions but the success and error responses are also defined by the same form.

### Controllers folder and API connection handler

> In order to keep logic simple to connect to the different APIs, the template uses a single axios instance in the httpClient.js. That allows us to create interceptors to define behaviors according to the responses of the service in only one part of the application.

Reviewer: ```suggestion
To keep the networking layer simple, the template uses a single Axios instance in the `httpClient.js`. It uses interceptors to define common side effects for the responses.
```

> When you need communication with a service you have to create an async function to manage the operation and grouping according to the kind of transaction inside a controller file, please keep all of those inside the controllers folder.

Reviewer 2: `controller's` folder

Reviewer 2: also, remove word `async` as many people may prefer to uses simple promises functions

> While the data transfer between the API and the app is working you must use the success and error actions that help you to catch the result of the operation. With this method you can see the hold process through the redux store. This is useful because you can create behaviours based on those states in a quick and simple way

Reviewer 2: see the `whole`? process

- `index.js`: Entry point of your application as per React-Native standards.

## Splash screen customization

To customize the splash screen (logo and background color) use the CLI provided in the [official docs](https://github.com/zoontek/react-native-bootsplash#assets-generation).

## Setup environments

### Using scripts from console

> The template allready has scripts to excecute the project calling an especific environment defined into the package.json file. Keep in mind that if you are going to create new `envs` you have to define the script to build the project propertly.

Reviewer: ```suggestion
The template already has scripts to execute the project calling a specific environment defined into the package.json file. Keep in mind that if you are going to create new `envs` you have to define the script to build the project properly.
```

Modify the environment variables files in root folder (`.env.development`, `.env.production` and `.env.staging`)

> 

Reviewer 2: @AGVasquez What do you think about adding another section here of how to create an archive for iOS and an APK for android? Maybe explaining the actions from xcode (select schema) and from console `./gradle ....`

## Setup environments

### Using scripts from console

> The template allready has scripts to excecute the project calling an especific environment defined into the package.json file. Keep in mind that if you are going to create new `envs` you have to define the script to build the project propertly.

Reviewer 2: `already`

- Click "Pre-actions", and under the plus sign select "New Run Script Action"
- Where it says "Type a script or drag a script file", type: `echo ".env.qa" > /tmp/envfile` replacing `.env.qa` with your file.

## Generate production version

> These are the steps to generate the .apk and .ipa files

Reviewer: ```suggestion
These are the steps to generate `.apk`, `.aab` and `.ipa` files
```
